.github/workflows/clang-tidy.yaml

```diff
@@ -0,0 +1,79 @@
+name: clang-tidy
+
+on:
+  push:
+    branches: [main]
+  pull_request:
+    branches: [main]
+
+jobs:
+  clang-tidy:
+    runs-on: ubuntu-24.04
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          submodules: recursive
+
+      - name: Setup Environment
+        env:
+          PROTOBUF_VERSION: '23.3'
+          ABSEIL_CPP_VERSION: '20230125.3'
+          CXX_STANDARD: '14'
+        run: |
+          sudo apt update -y
+          sudo apt install -y --no-install-recommends --no-install-suggests \
+          build-essential \
+          iwyu \
+          cmake \
+          ninja-build \
+          libssl-dev \
+          libcurl4-openssl-dev \
+          libprotobuf-dev \
+          protobuf-compiler \
+          libgmock-dev \
+          libgtest-dev \
+          libbenchmark-dev
+#          sudo -E ./ci/install_abseil.sh
+#          sudo -E ./ci/setup_grpc.sh -m -p protobuf -p abseil-cpp
+
+      - name: Prepare CMake
+        run: |
+          TOPDIR=$(pwd)
+          mkdir -p build && cd build
+          echo "Running cmake..."
+          CC="clang" CXX="clang++" cmake .. \
+          -DCMAKE_CXX_STANDARD=14 \
+          -DWITH_STL=CXX14 \
+          -DWITH_OTLP_HTTP=ON \
+          -DWITH_OTLP_FILE=ON \
+          -DWITH_PROMETHEUS=ON \
+          -DWITH_ZIPKIN=ON \
+          -DWITH_ELASTICSEARCH=ON \
+          -DWITH_OTLP_HTTP_COMPRESSION=ON \
+          -DWITH_EXAMPLES=ON \
+          -DWITH_EXAMPLES_HTTP=ON \
+          -DBUILD_W3CTRACECONTEXT_TEST=ON \
+          -DWITH_METRICS_EXEMPLAR_PREVIEW=ON \
+          -DWITH_ASYNC_EXPORT_PREVIEW=ON \
+          -DCMAKE_EXPORT_COMPILE_COMMANDS=ON \
+          -DCMAKE_CXX_CLANG_TIDY="clang-tidy-18"
```

Comment:
Not sure this is working properly, I don't see clang-tidy warnings in the log. Are you sure this is the correct binary name ?

There are build warnings from clang, yes, but the tidy warnings detected earlier are not there.

I suggest to add something like:

```
which clang-tidy
clang-tidy --version
```
in the worklow, to make sure clang-tidy is effectively installed.

This will help to debug CI.

Reply:
Yes, I'll add this. But, 

https://github.com/open-telemetry/opentelemetry-cpp/actions/runs/9909180749/job/27376671743

I can see the clang-tidy warnings for the job. Not sure why you are not able to view them.

Reply:
With iwyu, cmake runs iwyu first, and then runs the compiler (clang), to build a binary.

I suspect this works the same for clang-tidy, these warnings are from the compiler (clang).

For example, we know there are warnings about long/int64_t in the curl code, or warnings about exception in noexecpt methods ... where are they ?

Reply:
Locally, I get:

```
[  4%] Building CXX object api/test/context/propagation/CMakeFiles/composite_propagator_test.dir/composite_propagator_test.cc.o
/data/malff/CODE/MARC_GITHUB/opentelemetry-cpp/api/test/context/propagation/composite_propagator_test.cc:36:29: warning: C-style casts are discouraged; use constructor call syntax [google-readability-casting]
   36 |     auto it = headers_.find(std::string(key));
      |                             ^~~~~~~~~~~
      |                             std::string
/data/malff/CODE/MARC_GITHUB/opentelemetry-cpp/api/test/context/propagation/composite_propagator_test.cc:45:14: warning: C-style casts are discouraged; use constructor call syntax [google-readability-casting]
   45 |     headers_[std::string(key)] = std::string(value);
      |              ^~~~~~~~~~~
      |              std::string
/data/malff/CODE/MARC_GITHUB/opentelemetry-cpp/api/test/context/propagation/composite_propagator_test.cc:45:34: warning: C-style casts are discouraged; use constructor call syntax [google-readability-casting]
   45 |     headers_[std::string(key)] = std::string(value);
      |                                  ^~~~~~~~~~~
      |                                  std::string
[  4%] Linking CXX executable composite_propagator_test
```

so `-DCMAKE_CXX_CLANG_TIDY="clang-tidy"` is working.

Reply:
This is due to TOPDIR which I set which makefile detects. And removed it now. We can see the warnings.